### PR TITLE
Adds availability controller

### PR DESCRIPTION
**Pull request by GurkirtSingh**
### What this PR does (required):
- Creates controllers for availability model
- Creates middleware to fetch profile
- Creates validation function for new schedule request

### Any information needed to test this feature (required):
- Try creating new schedule by `/availability` using POST method
- Try fetching all the schedule by `/availability` using GET method
- Try fetching active schedule by '/availability/active' using GET method
- Try fetching schedule by id `/availability/:scheduleId` using GET method
- Try setting different active schedule by `availability/:scheduleId/activate` using PATCH method

### Any issues with the current functionality (optional):
- validation function needs to improve to only check `start` and `end` field if the `isAvailable` field is set to `true` on `days `object. And validate if `end` number is smaller then `start` number.

### CODE REVIEW

File: server/controllers/availability.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
+const asyncHandler = require("express-async-handler");
+const Profile = require("../models/Profile");
+const Schedule = require("../models/Schedule");
+
+exports.createSchedule = asyncHandler(async (req, res, next) => {
+  const profileId = req.profile.id;
+  const name = req.body.name;
+  const days = req.body.days;
```
Comment (sundayezeilo) [inline server/controllers/availability.js]: Nitpick: You can you a destructuring assignment here. You can also validate each and send a `400` if any is missing.

File: server/controllers/availability.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    res.status(201);
+    res.send("schedule is created");
+  } catch (err) {
+    res.status(500);
+    res.send(err);
+  }
+});
+// @route GET /availability/:scheduleId
+// @desc Get Schedule by Id
+// @access Private
```
Comment (sundayezeilo) [inline server/controllers/availability.js:39]: Good documentation

File: server/controllers/availability.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      res.send("schedule is not found!!");
+    }
+  } catch (err) {
+    res.status(500);
+    res.send(err);
+  }
+});
+// @route GET /availability
+// @desc Get list of all schedules
+// @access Private
```
Comment (sundayezeilo) [inline server/controllers/availability.js:55]: Nice

File: server/controllers/availability.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    res.send(schedules);
+  } catch (err) {
+    res.status(500);
+    res.send(err);
+  }
+});
+
+// @route GET /availability/active
+// @desc Get active schedule
+// @access Private
```
Comment (sundayezeilo) [inline server/controllers/availability.js:68]: :+1:

File: server/controllers/availability.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    res.send(schedule);
+  } catch (err) {
+    res.status(500);
+    res.send(err);
+  }
+});
+
+// @route PATCH /availability/:scheduleId/activate
+// @desc Set active schedule
+// @access Private
```
Comment (sundayezeilo) [inline server/controllers/availability.js:82]: :+1:

**Review by sundayezeilo — COMMENTED** [review]
Controller looks good. A few comments there.
Happy coding!

File: server/middleware/fetchProfile.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+const Profile = require("../models/Profile");
+
+const fetchProfile = async (req, res, next) => {
+  const userId = req.user.id;
+  if (!userId) {
+    return res.status(401).send("Not authorized");
+  }
+
+  const profile = await Profile.findOne({ userId: req.user.id });
```
Comment (rajivtitus) [inline server/middleware/fetchProfile.js]: Alternatively, attaching the profileId to the request object in the `protect` middleware after adding a ref to the Profile within the User model will things a lot easier for us 😄

File: server/controllers/availability.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
+const asyncHandler = require("express-async-handler");
+const Profile = require("../models/Profile");
+const Schedule = require("../models/Schedule");
+
+exports.createSchedule = asyncHandler(async (req, res, next) => {
+  if (req.profile === undefined) {
+    res.status(500);
```
Comment (rajivtitus) [inline server/controllers/availability.js]:
```suggestion
    res.status(404);
```

File: server/controllers/availability.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  const { name, days } = req.body;
+  if (!name) {
+    res.status(400);
+    throw new Error("schedule's name is required!!");
+  }
+  if (!days) {
+    res.status(400);
+    throw new Error("schedule's days object is required");
+  }
+  try {
```
Comment (rajivtitus) [inline server/controllers/availability.js]: We don't need a try/catch block as the `asyncHandler` will handle errors and forward it to the error middleware

Ditto where applicable

File: server/validate.js [hunk 9]
```diff
@@ -28,5 +28,45 @@ exports.validateLogin = [
     if (!errors.isEmpty())
       return res.status(400).json({ errors: errors.array() });
     next();
-  }
+  },
+];
+
+exports.validateScheduleSchema = [
```
Comment (rajivtitus) [inline server/validate.js:34]: 👍🏽

**Review by rajivtitus — APPROVED** [review]
Very good work with this! Please address comments before merging - conditionally approved

File: server/middleware/fetchProfile.js (hunk 6 above)
Comment (GurkirtSingh) [reply server/middleware/fetchProfile.js]: Why didn't I thought about it 😄

**Pull request merged**